### PR TITLE
Fix velon vbalance of at time

### PR DESCRIPTION
**Pull request by eric6311**
- fix vBalanceOfAtTime when target time is before lastPoint
- fix totalVBalanceAtTime target time is before lastPoint
- add more test cases for vBalance and totalVBalance

### CODE REVIEW

File: contracts/veLON.sol [hunk 1]
```diff
@@ -95,18 +95,27 @@ contract veLON is IveLON, ERC721, Ownable, ReentrancyGuard {
...
             return 0;
         } else {
-            Point memory lastPoint = userPointHistory[_tokenId][_epoch];
-            lastPoint.vBalance -= lastPoint.decliningRate * (int256(_t) - int256(lastPoint.ts));
-            if (lastPoint.vBalance < 0) {
-                lastPoint.vBalance = 0;
+            // search for the userEpoch where _t is bigger then userPointHistory[_tokenId][_epoch].ts,
+            // but smaller then userPointHistory[_tokenId][_epoch+1].ts
+            for (uint256 i = _epoch; i > 0; i--) {
+                if (userPointHistory[_tokenId][i].ts <= _t) {
```
Comment (charlesjhongc) [inline contracts/veLON.sol:107]: `userPointHistory[_tokenId][i]` can be stored into memory first, to save some gas.

File: contracts/test/veLON/vBalance.t.sol [hunk 2]
```diff
@@ -189,6 +189,75 @@ contract TestVeLONBalance is TestVeLON {
...
+        assertEq(veLon.userPointEpoch(tokenId), 2);
+
+        // Continue fast forward 1 week, so now is the third week(2 weeks passed since createLock)
+        // Query the vBalanceOfAtime in first week (the time lock was created)
+        // The traget time is before lastPoint.ts(the time lock was extended)
+        vm.warp(block.timestamp + 1 weeks);
+        vm.roll(block.number + 1);
+        uint256 initialBalance = _initialvBalance(DEFAULT_STAKE_AMOUNT, 2 weeks);
+        assertEq(veLon.userPointEpoch(tokenId), 2);
+        assertEq(veLon.vBalanceOfAtTime(tokenId, block.timestamp - 2 weeks), initialBalance);
```
Comment (charlesjhongc) [inline contracts/test/veLON/vBalance.t.sol:249]: How about check balance at `ts` other than the begining? For example a `ts` between 2 user points.

File: contracts/veLON.sol [hunk 3]
```diff
@@ -178,6 +187,26 @@ contract veLON is IveLON, ERC721, Ownable, ReentrancyGuard {
...
+                break;
+            }
+            uint256 _mid = _min.add(_max).add(1).div(2);
+            if (poolPointHistory[_mid].ts <= _t) {
+                _min = _mid;
+            } else {
+                _max = _mid.sub(1);
+            }
+        }
+        return _min;
```
Comment (zeqing-guo) [inline contracts/veLON.sol:207]: Should we consider the situation that there are several updates in one block?

File: contracts/veLON.sol (hunk 1 above)
Comment (eric6311) [reply contracts/veLON.sol]: refactor in https://github.com/consenlabs/tokenlon-contracts/pull/81/commits/3d00ba0be76a151013d564cb2a1ee6dd9aba7cac

File: contracts/test/veLON/vBalance.t.sol (hunk 2 above)
Comment (eric6311) [reply contracts/test/veLON/vBalance.t.sol]: change in [ffd08d6](https://github.com/consenlabs/tokenlon-contracts/pull/81/commits/ffd08d65e0fe8a5308e3b267720d93e4543d6c45)

File: contracts/test/veLON/vBalance.t.sol [hunk 4]
```diff
@@ -189,6 +189,75 @@ contract TestVeLONBalance is TestVeLON {
...
+        uint256 aliceBalance = _initialvBalance(DEFAULT_STAKE_AMOUNT, 2 weeks);
+
+        assertEq(veLon.epoch(), 1);
+        assertEq(veLon.vBalanceOfAtBlk(aliceTokenId, block.number), aliceBalance);
+
+        // fast forward 1 week and second user createlock
+        vm.warp(block.timestamp + 1 weeks);
+        vm.roll(block.number + 1);
+        uint256 bobTokenId = _stakeAndValidate(bob, DEFAULT_STAKE_AMOUNT, 2 weeks);
+        uint256 bobBalance = _initialvBalance(DEFAULT_STAKE_AMOUNT, 2 weeks);
```
[the remark this answers is not in the document]
Comment (eric6311) [reply contracts/test/veLON/vBalance.t.sol]: remove in https://github.com/consenlabs/tokenlon-contracts/pull/81/commits/ffd08d65e0fe8a5308e3b267720d93e4543d6c45

File: contracts/test/veLON/vBalance.t.sol (hunk 2 above)
Comment (eric6311) [reply contracts/test/veLON/vBalance.t.sol]: change in https://github.com/consenlabs/tokenlon-contracts/pull/81/commits/ffd08d65e0fe8a5308e3b267720d93e4543d6c45

File: contracts/veLON.sol [hunk 5]
```diff
@@ -463,14 +492,14 @@ contract veLON is IveLON, ERC721, Ownable, ReentrancyGuard {
...
         // (approximately, for *At methods) and save them
         // as we cannot figure that out exactly from inside the contract
-        Point memory storedPoolLastPoint = poolLastPoint;
+        Point memory storedPoolLastPoint = Point({ vBalance: 0, decliningRate: 0, ts: poolLastPoint.ts, blk: poolLastPoint.blk });
 
-        // blockRate = dBlock/dTime
+        // blockRate = dTime/dBlock
         // If last point is already recorded in this block, blockRate=0
         // But that's ok because we know the block in such case
         uint256 blockRate = 0;
```
Comment (charlesjhongc) [inline contracts/veLON.sol:500]: Rename as `avgBlkTime`.

File: contracts/veLON.sol [hunk 6]
```diff
@@ -500,7 +529,11 @@ contract veLON is IveLON, ERC721, Ownable, ReentrancyGuard {
 
             // update ts and block (approximately block number)
             poolLastPoint.ts = nextPointTs;
-            poolLastPoint.blk = storedPoolLastPoint.blk + (blockRate * (nextPointTs - storedPoolLastPoint.ts)) / MULTIPLIER;
+            if (blockRate > 0) {
+                poolLastPoint.blk = storedPoolLastPoint.blk + ((nextPointTs - storedPoolLastPoint.ts) * MULTIPLIER) / blockRate;
```
Comment (charlesjhongc) [inline contracts/veLON.sol:533]: In this way, `MULTIPLIER` may be dropped.

Comment (eric6311) [reply contracts/veLON.sol]: remove in https://github.com/consenlabs/tokenlon-contracts/pull/81/commits/ef590c464974ef594ff79e9bae2d6d986f5b114d

File: contracts/veLON.sol (hunk 5 above)
Comment (eric6311) [reply contracts/veLON.sol]: rename in https://github.com/consenlabs/tokenlon-contracts/pull/81/commits/ef590c464974ef594ff79e9bae2d6d986f5b114d

**Review by charlesjhongc — APPROVED** [review]

**Pull request merged**